tests/framework/extensions/nodes/ec2/ec2.go

```diff
@@ -82,82 +92,66 @@ func CreateNodes(client *rancher.Client, numOfInstances int, numOfWinInstances i
 			InstanceIds: listOfInstanceIds,
 		})
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		//wait until instance status is ok
 		err = ec2Client.SVC.WaitUntilInstanceStatusOk(&ec2.DescribeInstanceStatusInput{
 			InstanceIds: listOfInstanceIds,
 		})
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
-		// describe instance to get attributes=
+		// describe instance to get attributes
 		describe, err := ec2Client.SVC.DescribeInstances(&ec2.DescribeInstancesInput{
 			InstanceIds: listOfInstanceIds,
 		})
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		readyInstances := describe.Reservations[0].Instances
 
-		sshKey, err := nodes.GetSSHKey(config.AWSSSHKeyName)
+		sshKey, err := nodes.GetSSHKey(reservationConfigs[i].AWSSSHKeyName)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		for _, readyInstance := range readyInstances {
-
-			if multiconfig {
-				if config.AWSAMI == ec2Client.ClientConfig.AWSEC2Config[0].AWSAMI &&
-					config.AWSUser == ec2Client.ClientConfig.AWSEC2Config[0].AWSUser {
-					ec2Node := &nodes.Node{
-						NodeID:          *readyInstance.InstanceId,
-						PublicIPAddress: *readyInstance.PublicIpAddress,
-						SSHUser:         config.AWSUser,
-						SSHKey:          sshKey,
-					}
-					ec2Nodes = append(ec2Nodes, ec2Node)
-				}
-
-				if config.AWSAMI == ec2Client.ClientConfig.AWSEC2Config[1].AWSAMI &&
-					config.AWSUser == ec2Client.ClientConfig.AWSEC2Config[1].AWSUser {
-					ec2Node2 := &nodes.Node{
-						NodeID:          *readyInstance.InstanceId,
-						PublicIPAddress: *readyInstance.PublicIpAddress,
-						SSHUser:         config.AWSUser,
-						SSHKey:          sshKey,
-					}
-					winEC2Nodes = append(winEC2Nodes, ec2Node2)
-				}
-			} else {
-				ec2Node := &nodes.Node{
-					NodeID:          *readyInstance.InstanceId,
-					PublicIPAddress: *readyInstance.PublicIpAddress,
-					SSHUser:         config.AWSUser,
-					SSHKey:          sshKey,
-				}
-				ec2Nodes = append(ec2Nodes, ec2Node)
-				winEC2Nodes = nil
+			ec2Node := &nodes.Node{
+				NodeID:           *readyInstance.InstanceId,
+				PublicIPAddress:  *readyInstance.PublicIpAddress,
+				PrivateIPAddress: *readyInstance.PrivateIpAddress,
+				SSHUser:          reservationConfigs[i].AWSUser,
+				SSHKey:           sshKey,
 			}
+			// re-reverse the list so that the order is corrected
+			ec2Nodes = append([]*nodes.Node{ec2Node}, ec2Nodes...)
 		}
+	}
 
-		client.Session.RegisterCleanupFunc(func() error {
-			return DeleteNodes(client, ec2Nodes, winEC2Nodes, multiconfig)
-		})
+	client.Session.RegisterCleanupFunc(func() error {
+		return DeleteNodes(client, ec2Nodes)
+	})
 
-		if !multiconfig {
-			break
+	return ec2Nodes, nil
+}
+
+// MatchRoleToConfig matches the role of nodesAndRoles to the ec2Config that allows this role.
+func MatchRoleToConfig(poolRole string, ec2Configs []rancherEc2.AWSEC2Config) rancherEc2.AWSEC2Config {
```

Comment:
I agree as well, that way also on line 23 `reservationConfigs = append(reservationConfigs, &config)` can just accept the pointer. Should be an easy update. No Jenkins run needed.

Reply:
I can't use a [pointer for the array](https://stackoverflow.com/questions/62095562/indexing-with-pointers-how-pointer-works-with-slices) but I can return a pointer instead. Updating the PR now.

Reply:
if you mean the other way around, I don't know how to convert the []awsEc2Configs to []*awsEc2Configs in a clean way

Reply:
Agreed, like you mentioned Caleb, seems the only way is changing the configuration cause it uses directly from the configuration. I think it would make this PR's scope wider unnecessarily. So the current change makes sense to me, thank you!